src/schoology.js

```diff
@@ -0,0 +1,381 @@
+    accessToken = {},
+    fxs = {},
+  }) {
+    this.hostedUrl = hostedUrl;
+    this.redirectUri = redirectUri;
+    this.clientId = clientId;
+    this.clientSecret = clientSecret;
+    this.userId = userId;
+    this.cacheRequestToken = fxs.cacheRequestToken || (() => {});
+    this.getUserAccessToken = fxs.getAccessToken || (() => {});
+    this.setUserAccessToken = fxs.setAccessToken || (() => {});
+
+    this.oAuth = new OAuth( {
+      consumerKey: this.clientId, 
+      consumerSecret: this.clientSecret, 
+      apiBase: 'https://api.schoology.com', 
+      authRealm: 'Schoology API',
+      signatureMethod: 'PLAINTEXT',
+      nonceLength: 16,
+      requestToken,
+      accessToken,
+      handleError: this.handleError
+    });
+  }
+
+  async build() {
+    if (!this.accessToken && !this.refreshToken) {
+      await this.getTokensFromUser();
+    }
+    return this;
+  }
+
+  /**
+   * Returns a URL used to initiate the authorization process with schoology and fetch
+   * the authorization code
+   */
+   async getAuthorizationURL(options = {}) {
+    try {
+      const result = await this.oAuth.getRequestTokens('/v1/oauth/request_token');
+      const tokenData = result.response;
+
+      await this.cacheRequestToken(tokenData);
+      
+      return OAuth.makeURL( this.hostedUrl, '/oauth/authorize', {
+        'oauth_token': tokenData.token,
+        'oauth_callback': this.redirectUri,
+      } )
+    } catch ( error ) {
+      this.handleError(error)
+    }
+  }
+
+  /**
+   * Fetches the access and refresh tokens for a valid authorization code
+   */
+   async getAccessTokens() {
+    try {
+      const result = await this.oAuth.getAccessTokens('/v1/oauth/access_token');
+      const tokenData = result.response;
+      
+      await this.setUserAccessToken( tokenData );
+      
+      return tokenData;
+    } catch ( error ) {
+      this.handleError(error)
+    }
+  }
+
+  async getProfile() {
+    try {
+      const resp = await this.makeRequest({
+        url: '/v1/users/me',
+        method: 'GET',
+        headers: {
+          'Content-Type': 'application/json',
+        },
+      });
+      this.schoologyUserId = resp.data.id;
+      return resp.data;
+    } catch(err) {
+      throw new LMSError('Unable to fetch user profile', 'schoology.USER_PROFILE_ERROR', {
+        userId: this.userId
+      });
+    }
+  }
+
+  async getTokensFromUser() {
+    try {
+      const { accessToken, refreshToken, info } = await this.getUserToken(this.userId);
+      this.accessToken = accessToken;
+      this.refreshToken = refreshToken;
+      this.schoologyUserId = info.id;
+    } catch (err) {
+      throw new LMSError('Unable to fetch tokens from user', 'schoology.TOKEN_FETCH_ERROR', {
+        userId: this.userId,
+        message: err.message,
+      });
+    }
+  }
+
+  /**
+   * Handles some schoology API errors
+   */
+  handleError(err, code) {
+    if (err.response) {
+      switch (err.response.status) {
```

Comment:
We need to handle the case where the user's oauth token expires. You can have a look at how canvas handles it, ideally we'd try to refresh it automatically a few times before throwing an error (which would take them through the oauth flow again)